src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -32,7 +32,19 @@ public sealed partial class JsonSerializerOptions
         /// so using fresh default instances every time one is needed can result in redundant recomputation of converters.
         /// This property provides a shared instance that can be consumed by any number of components without necessitating any converter recomputation.
         /// </remarks>
-        public static JsonSerializerOptions Default { get; } = CreateDefaultImmutableInstance();
+        public static JsonSerializerOptions Default
+        {
+            [RequiresUnreferencedCode(JsonSerializer.SerializationUnreferencedCodeMessage)]
+            [RequiresDynamicCode(JsonSerializer.SerializationRequiresDynamicCodeMessage)]
+            get
+            {
+                JsonSerializerOptions defaultInstance = DefaultInstance;
+                defaultInstance._typeInfoResolver ??= DefaultJsonTypeInfoResolver.RootDefaultInstance();
+                return defaultInstance;
+            }
+        }
+
+        internal static JsonSerializerOptions DefaultInstance { get; } = new JsonSerializerOptions { IsLockedInstance = true };
```

Comment:
Internal consumers of this will see a locked JsonSerializerOptions, but if those consumers are concurrently used with public consumers of Default, they could see its _typeInfoResolver transition from null to non-null.  Is that ok?

Reply:
It is a bit of a wrinkle, but the current consumers will never leak that instance to user code since it's being passed to hardcoded converters, e.g.

https://github.com/dotnet/runtime/blob/3586889c863a8ece6a4869d3367a0017c49b8d82/src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs#L99-L107

or

https://github.com/dotnet/runtime/blob/3586889c863a8ece6a4869d3367a0017c49b8d82/src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValueTrimmable.cs#L37-L46

which is an internal type being used for methods like the following

https://github.com/dotnet/runtime/blob/8cb4e934757b959bc1b38aecaa764e21a47ec85e/src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValue.CreateOverloads.cs#L16

I'm not sure how we might prevent accidental misuse of `DefaultInstance` in the future though.

Reply:
> It is a bit of a wrinkle, but the current consumers will never leak that instance to user code since it's being passed to hardcoded converters

I don't understand: the public JsonSerializerOptions.Default is returning this instance.  Can you elaborate?

Reply:
Using a dedicated options singleton in JsonNode might just work.

Reply:
FWIW I changed `JsonNode` to use its own instance -- `JsonSerializerOptions.Default` is now guaranteed to be immutable on construction.